consensus/src/sync/live/diff_queue/mod.rs

```diff
@@ -213,49 +222,30 @@ impl<N: Network> Stream for DiffQueue<N> {
     type Item = QueuedDiff<N>;
 
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-        let mut block_queue_done = false;
-
         // Receive blocks from BlockQueue.
-        loop {
-            match self.block_queue.poll_next_unpin(cx) {
-                Poll::Ready(Some(block)) => {
-                    if !self.diff_needed {
-                        return Poll::Ready(Some(match block {
-                            QueuedBlock::Head(block) => QueuedDiff::Head(block, None),
-                            QueuedBlock::Buffered(blocks) => QueuedDiff::Buffered(
-                                blocks.into_iter().map(|block| (block, None)).collect(),
-                            ),
-                            QueuedBlock::Missing(blocks) => QueuedDiff::Missing(
-                                blocks.into_iter().map(|block| (block, None)).collect(),
-                            ),
-                            QueuedBlock::TooFarAhead(peer_id) => QueuedDiff::TooFarAhead(peer_id),
-                            QueuedBlock::TooFarBehind(peer_id) => QueuedDiff::TooFarBehind(peer_id),
-                        }));
-                    }
-
+        while let Poll::Ready(block) = self.block_queue.poll_next_unpin(cx) {
+            match (block, self.diff_needed) {
+                (Some(block), true) => {
                     let get_diff = self.diff_request_component.request_diff();
                     self.diffs
                         .push_back(Box::pin(augment_block(block, get_diff)));
                 }
-                Poll::Ready(None) => {
-                    block_queue_done = true;
-                    break;
+                (Some(block), false) => {
+                    self.diffs.push_back(Box::pin(future::ready(Ok(
+                        QueuedDiff::from_block_no_diff(block),
+                    ))));
                 }
-                Poll::Pending => break,
+                (None, ..) => return Poll::Ready(None),
```

Comment:
```suggestion
                // Immediately quit when the block queue stream has ended, not
                // waiting for any pending diff requests.
                (None, ..) => return Poll::Ready(None),
```
One last nit: Could you add a comment saying that we explicitly end the stream if the block queue stream has ended, not waiting for the diff requests (just to make sure that it's clear that this is intended)?